src/EditorFeatures/Core/Diagnostics/AbstractPushOrPullDiagnosticsTaggerProvider.PushDiagnosticsTaggerProvider.cs

```diff
@@ -0,0 +1,248 @@
+            // cause TryGetText to fail.  However, that's ok.  In that case, if we do need to tag this
+            // document, we'll just use the current editor snapshot.  If that's the same, then the tags
+            // will be hte same.  If it is different, we'll eventually hear about the new diagnostics 
+            // for it and we'll reach our fixed point.
+            if (document != null && document.IsOpen())
+            {
+                // This should always be fast since the document is open.
+                var sourceText = document.State.GetTextSynchronously(cancellationToken: default);
+                snapshot = sourceText.FindCorrespondingEditorTextSnapshot();
+                if (snapshot != null)
+                {
+                    _diagnosticIdToTextSnapshot.GetValue(e.Id, _ => snapshot);
+                }
+            }
+        }
+
+        protected sealed override TaggerDelay EventChangeDelay => TaggerDelay.Short;
+        protected sealed override TaggerDelay AddedTagNotificationDelay => TaggerDelay.OnIdle;
+
+        protected sealed override ITaggerEventSource CreateEventSource(ITextView? textView, ITextBuffer subjectBuffer)
+            => CreateEventSourceWorker(subjectBuffer, _diagnosticService);
+
+        protected override Task ProduceTagsAsync(
+            TaggerContext<TTag> context, DocumentSnapshotSpan spanToTag, int? caretPosition, CancellationToken cancellationToken)
+        {
+            return ProduceTagsAsync(context, spanToTag, cancellationToken);
+        }
+
+        private async Task ProduceTagsAsync(
+            TaggerContext<TTag> context, DocumentSnapshotSpan spanToTag, CancellationToken cancellationToken)
+        {
+            if (!_callback.IsEnabled)
+                return;
+
+            var diagnosticMode = GlobalOptions.GetDiagnosticMode(InternalDiagnosticsOptions.NormalDiagnosticMode);
+            if (!_callback.SupportsDiagnosticMode(diagnosticMode))
+                return;
+
+            var document = spanToTag.Document;
+            if (document == null)
+                return;
+
+            var snapshot = spanToTag.SnapshotSpan.Snapshot;
+            var workspace = document.Project.Solution.Workspace;
+
+            // See if we've marked any spans as those we want to suppress diagnostics for.
+            // This can happen for buffers used in the preview workspace where some feature
+            // is generating code that it doesn't want errors shown for.
+            var buffer = snapshot.TextBuffer;
+            var suppressedDiagnosticsSpans = (NormalizedSnapshotSpanCollection?)null;
+            buffer?.Properties.TryGetProperty(PredefinedPreviewTaggerKeys.SuppressDiagnosticsSpansKey, out suppressedDiagnosticsSpans);
+
+            var buckets = diagnosticMode switch
+            {
+                DiagnosticMode.Pull => _diagnosticService.GetPullDiagnosticBuckets(workspace, document.Project.Id, document.Id, diagnosticMode, cancellationToken),
```

Comment:
Shouldn't this always be using push?

Reply:
i will comment, but as per our discussion, no.  The DiagnosticMode is whether we are on Lsp or not (and i'll rename accordingly).  We can be in lsp pull mode, but still be on tagging push mode (e.g. inline-diagnostics in tagging push mode).